Schedule actions at a given time

alexandredantas opened the pull request:
Adding functions enabling an effect to run only at a given point in time.
The functions `exec` and `await` call each other to establish when an effect must run.

The `exec` function calls `foldM` to execute the effect and retrieve its result. At this point, it gives control to `await`

The `await` function awaits when schedule signal when it's ready to run. Schedule will always receive. Schedules for this always receive unit as input because they don't need effect to run to decide when to run
When schedule is ready to run, `await` return control to `exec`, restarting the process

alexandredantas commented:
Two points on this PR: 
  - i've not added unit tests because i never wrote specs before. I would really appreciate some help to write the tests.
  - when building microsite, i got a failure related to jekyll. Was it my fault?

alexandredantas commented on core/shared/src/main/scala/scalaz/zio/ZSchedule.scala:
```diff
@@ -71,7 +75,7 @@ trait ZSchedule[-R, -A, +B] extends Serializable { self =>
         case a :: as =>
           self.update(a, s).flatMap {
-            case ZSchedule.Decision(cont, delay, s, finish) =>
+            case ZSchedule.Decision(cont, _, delay, s, finish) =>
```
Here i want a suggestion: i'm not considering readiness on run. Should I? If yes, what should be the behavior? Discard results when not ready? Sleep until next readiness?

jdegoes commented:
@alexandredantas I think this is a useful feature to have.

However, I think we can do it in a slightly different way that has a cleaner design: adding a method into `Scheduler` that allows scheduling something at a particular instant.

Then, potentially, every step could return either a delay (relative time) or an instant (absolute time). I have to think this through, but I think this has all the structure required of `Schedule` and the existing combinators.

Do you think that would work?

alexandredantas commented:
@jdegoes make total sense. convert the point at time to a delay and effects would sleep until reaching it. i'll rework on that. thanks for the tip!

ps: can you help pointing out which properties i should test? :)

jdegoes commented:
@alexandredantas The main thing we need to ensure is that invariants around `&&` and `||` are respected. For example: semantics of `&&` are to choose the larger of the two delays; and `||`, to choose the smaller of the two delays.

Things will be easier if we introduce a type:

```scala
sealed trait Delay
object Delay {
  final case class Relative(duration: Duration)
  final case class Absolute(time: Instant) // Cannot use Instant for Scala.js???
}
```

Now our problem is comparing them to see which one comes first requires an effect, because we have to use the `Clock`, so our comparison methods will change:

```scala
sealed trait Delay {
  def < (that: Delay): ZIO[Clock, Nothing, Boolean] = ???
  def <= (that: Delay): ZIO[Clock, Nothing, Boolean] = ???
  def > (that: Delay): ZIO[Clock, Nothing, Boolean] = ???
  def >= (that: Delay): ZIO[Clock, Nothing, Boolean] = ???
  def === (that: Delay): ZIO[Clock, Nothing, Boolean] = ???
}
```

Direct comparison on `Duration` happens for `&&` and `||` methods. It's not desirable to make them do the comparison (instead we should push that all the way to `ZIO#retry` / `ZIO#repeat`). So we have to change that by tweaking `combineWith`, using some new data structure, which can describe comparisons without actually making them:

```scala
sealed trait DelayComparison { self =>
  def choose: ZIO[Clock, Nothing, Delay] = 
    self match {
      ??? // Uses effectful comparison methods on `Delay`
    }
}
object DelayComparison {
  final case class Choose(value: Delay) extends DelayComparison
  final case class Min(l: DelayComparison, r: DelayComparison) extends DelayComparison
  final case class Max(l: DelayComparison, r: DelayComparison) extends DelayComparison
}
```

Now `combineWith` will not take a function `(Duration, Duration) => Duration`, but rather, a `(Delay, Delay) => DelayComparison`. So now `Clock` dependency (pulled in by comparing `Delay`) will not contaminate the `Schedule` methods, but rather be pushed into `ZIO#retry` / `ZIO#repeat`, which will call the `DelayComparison#choose` method.

At this point all existing schedules will be updated to use `Delay.Relative` (with helper: `Delay.relative`), since that's what they use now. And some new schedules can use `Delay.Absolute`, e.g. `Schedule.at(...)`.

alexandredantas commented:
@jdegoes Wow!! Such a very nice solution!! I'll come up with something following your suggestion

alexandredantas commented:
@jdegoes can you tell me if I'm on the right direction? When I changed Decision.combineWith to (Delay, Delay) => DelayComparison, it was required to touch every other schedule implementation.
Besides, combineWith take Direction.delay as arguments and I have changed it to be DelayComparison, not allowing to apply to the function :(

What did I do wrong?

alexandredantas commented:
@jdegoes After banging my head against wall sometimes, I've came up with this implementation. Adjusted tests accordingly and at least build is passing :smile: 
If things looks good, I want to place a schedule capable of considering only a point in time. However, scalajs doesn't support `LocalDateTime`. Is there something else I could use to perform date calculations?

jdegoes commented:
@alexandredantas After looking at some of the contortions you had to go through, I can see now what I should have seen all alone: `Delay` and `DelayComparison` need to be fused into a single structure, called `Delay`.


```scala
sealed trait Delay {
  def run: ZIO[Clock, Nothing, Duration, Instant]
}
object Delay {
  val none: Delay = Relative(Duration.zero)
  // Etc.

  final case class Relative(duration: Duration) extends Delay
  final case class Absolute(instant: Instant) extends Delay // Scala.js compat?
  final case class Min(left: Delay, right: Delay) extends Delay
  final case class Max(left: Delay, right: Delay) extends Delay
|
```

Now functions can be `(Delay, Delay) => Delay` and composition will be vastly superior. Should clean up a lot of code, too!

alexandredantas commented:
@jdegoes :smile: Sorry for noobiness. I should have tried that first instead implementing them separately. I'll clean up things and do as you suggested. After all, do things look good? Do you see any major mistake?

jdegoes commented:
@alexandredantas I think a lot of the gnarlier code will be clean after this change. The overall direction of changes seems correct to me. There may be some edge case lodging that needs more careful review when the bigger changes have been made, but you're definitely on the right track! 😄

jdegoes commented:
@alexandredantas Let me know if you need a hand. 😄

alexandredantas commented:
@jdegoes Tks!! :smile: I wasn't able to work on this in past days but I'll get back soon

alexandredantas commented:
@jdegoes Finished fusing DelayComparison and Delay. Code is a lot cleaner now. It allowed to compute delay directly instead returning a tuple as result of schedule effect, decoupling implementations to know about relative and absolute semantics: just need to sleep for the given duration. This design actually is a lot better, preventing clock drifts and making cpu time optimized.

Now I'm solving some issues with dotty inside schedule, ok?

alexandredantas commented:
@jdegoes Instant, LocalDate and LocalTime are suported in scalajs through this library: https://github.com/scala-js/scala-js-java-time
So I don't think is a big deal using these classes for now. What do you think?

jdegoes commented on core/jvm/src/test/scala/scalaz/zio/RetrySpec.scala:
```diff
@@ -119,29 +124,50 @@ class RetrySpec(implicit ee: org.specs2.concurrent.ExecutionEnv) extends TestRun
+
+    val results = scheduled.zip(expected).map {
+      case (r1, r2) =>
+        unsafeRun(for {
```
Push this unsafe run to the top, by making a top-level for comprehension. Each test should have just 1 unsafeRun.

jdegoes commented on core/jvm/src/test/scala/scalaz/zio/RetrySpec.scala:
```diff
@@ -119,29 +124,50 @@ class RetrySpec(implicit ee: org.specs2.concurrent.ExecutionEnv) extends TestRun
-    val scheduled: List[(Duration, Int)] = unsafeRun(
+    val schedule: ZSchedule[Random, Int, Int] =
+      Schedule.recurs(5).delayed(_ => 500.millis.relative).jittered
+    val scheduled: List[(Delay, Int)] = unsafeRun(
```
Push this unsafe run to the top, by making a top-level for comprehension. Each test should have just 1 unsafeRun.

jdegoes commented on core/jvm/src/test/scala/scalaz/zio/RetrySpec.scala:
```diff
@@ -119,29 +124,50 @@ class RetrySpec(implicit ee: org.specs2.concurrent.ExecutionEnv) extends TestRun
-    val scheduled: List[(Duration, Int)] = unsafeRun(
+    val schedule: ZSchedule[Random, Int, Int] =
+      Schedule.recurs(5).delayed(_ => 500.millis.relative).jittered(2, 4)
+    val scheduled: List[(Delay, Int)] = unsafeRun(
```
Push this unsafe run to the top, by making a top-level for comprehension. Each test should have just 1 unsafeRun.

jdegoes commented on core/jvm/src/test/scala/scalaz/zio/RetrySpec.scala:
```diff
@@ -119,29 +124,50 @@ class RetrySpec(implicit ee: org.specs2.concurrent.ExecutionEnv) extends TestRun
-    scheduled must_=== expected
+    val expected = List(1, 2, 3, 4, 5).map((1500.millis.relative, _))
+
+    val results = scheduled.zip(expected).map {
```
Push this unsafe run to the top, by making a top-level for comprehension. Each test should have just 1 unsafeRun.

jdegoes commented on core/jvm/src/test/scala/scalaz/zio/RetrySpec.scala:
```diff
@@ -151,8 +177,30 @@ class RetrySpec(implicit ee: org.specs2.concurrent.ExecutionEnv) extends TestRun
+      expected._1,
+      expected._2.map(
+        r1 =>
+          unsafeRun(for {
```
Push this (and the other 2) unsafe runs to the top, by making a top-level for comprehension. Each test should have just 1 unsafeRun.

jdegoes commented on core/jvm/src/test/scala/scalaz/zio/RetrySpec.scala:
```diff
@@ -178,15 +226,47 @@ class RetrySpec(implicit ee: org.specs2.concurrent.ExecutionEnv) extends TestRun
+      retried._1,
+      retried._2.map(
+        r1 =>
+          unsafeRun(for {
```
Same thing here.

jdegoes commented on core/shared/src/main/scala/scalaz/zio/ZSchedule.scala:
```diff
@@ -197,7 +203,7 @@ trait ZSchedule[-R, -A, +B] extends Serializable { self =>
    */
   final def &&[R1 <: R, A1 <: A, C](that: ZSchedule[R1, A1, C]): ZSchedule[R1, A1, (B, C)] =
-    combineWith(that)(_ && _, _ max _)
+    combineWith(that)(_ && _, (a, b) => Max(a, b))
```
I'd add `max` and `min` functions to `Delay` so you don't have to change this code.

jdegoes commented on core/shared/src/main/scala/scalaz/zio/ZSchedule.scala:
```diff
@@ -239,7 +245,7 @@ trait ZSchedule[-R, -A, +B] extends Serializable { self =>
    */
   final def ||[R1 <: R, A1 <: A, C](that: ZSchedule[R1, A1, C]): ZSchedule[R1, A1, (B, C)] =
-    combineWith(that)(_ || _, _ min _)
+    combineWith(that)(_ || _, (a, b) => Min(a, b))
```
I'd add `max` and `min` functions to `Delay` so you don't have to change this code.

jdegoes commented on core/shared/src/main/scala/scalaz/zio/ZSchedule.scala:
```diff
@@ -435,7 +441,7 @@ trait ZSchedule[-R, -A, +B] extends Serializable { self =>
         self.update(a, s._1).flatMap { step1 =>
           that.update(step1.finish(), s._2).map { step2 =>
-            step1.combineWith(step2)(_ && _, _ + _).rightMap(_._2)
+            step1.combineWith(step2)(_ && _, (a, b) => Sum(a, b)).rightMap(_._2)
```
I'd add `+` functions to `Delay` so you don't have to change this code.

jdegoes commented on core/shared/src/main/scala/scalaz/zio/ZSchedule.scala:
```diff
@@ -482,7 +488,9 @@ trait ZSchedule[-R, -A, +B] extends Serializable { self =>
-        self.update(a._1, s._1).zipWith(that.update(a._2, s._2))(_.combineWith(_)(_ && _, _ max _))
+        self
+          .update(a._1, s._1)
+          .zipWith(that.update(a._2, s._2))(_.combineWith(_)(_ && _, (a, b) => Max(a, b)))
```
I'd add `max` function to `Delay` so you don't have to change this code.

jdegoes commented on core/shared/src/main/scala/scalaz/zio/ZSchedule.scala:
```diff
@@ -564,9 +575,9 @@ trait Schedule_Functions extends Serializable {
+  final def delayed[R: ConformsR, A](s: ZSchedule[R, A, Delay]): ZSchedule[R, A, Delay] = {
     val delayed = s.modifyDelay((b, d) => IO.succeed(b + d))
-    delayed.reconsider((_, step) => step.copy(finish = () => step.delay)) // TODO: Dotty doesn't infer this properly
+    delayed.reconsider((_, step) => step.copy(finish = () => step.delay))
```
Please add the comment about Dotty, it's how we are tracking Dotty type inference regressions.

jdegoes commented on core/shared/src/main/scala/scalaz/zio/ZSchedule.scala:
```diff
@@ -654,33 +665,35 @@ trait Schedule_Functions extends Serializable {
    */
   final def spaced(interval: Duration): Schedule[Any, Int] =
-    forever.delayed(_ + interval)
+    forever.delayed(d => d + relative(interval))
```
`_ + relative(interval)`

jdegoes commented on core/shared/src/main/scala/scalaz/zio/delay/Delay.scala:
```diff
@@ -0,0 +1,38 @@
+  }
+
+  def *(factor: Double) = TimesFactor(self, factor)
+  def +(delay: Delay)   = Sum(self, delay)
```
Excellent work, this is looking really great! A few minor suggestions:

1. Rename `TimesFactor` to `Scale`
2. Add `min`, `max`

jdegoes commented on core/shared/src/main/scala/scalaz/zio/delay/Delay.scala:
```diff
@@ -0,0 +1,38 @@
+  final case class Sum(l: Delay, r: Delay)               extends Delay
+
+  val none: Delay                     = Relative(Duration.Zero)
+  final def relative(delay: Duration) = Relative(delay)
```
I'd add a helper `absolute`, too, that way direct constructors need not be used.

jdegoes commented on core/shared/src/main/scala/scalaz/zio/delay/DelaySyntax.scala:
```diff
@@ -0,0 +1,7 @@
+
+import scalaz.zio.duration.Duration
+
+class DelaySyntax(duration: Duration) {
```
Consider adding this as a method to `Duration` instead.

jdegoes commented on core/shared/src/main/scala/scalaz/zio/ZSchedule.scala:
```diff
@@ -654,33 +665,35 @@ trait Schedule_Functions extends Serializable {
-    delayed(unfold[(Duration, Duration)]((Duration.Zero, one)) {
-      case (a1, a2) => (a2, a1 + a2)
-    }.map(_._1))
+  final def fibonacci(one: Duration): Schedule[Any, Delay] =
```
Try to return `Duration` here instead of `Delay`, by running the `Delay`. We might need to add `mapM` to schedule to do that easily.

jdegoes commented on core/shared/src/main/scala/scalaz/zio/ZSchedule.scala:
```diff
@@ -654,33 +665,35 @@ trait Schedule_Functions extends Serializable {
    */
-  final def linear(base: Duration): Schedule[Any, Duration] =
-    delayed(forever.map(i => base * i.doubleValue()))
+  final def linear(base: Duration): Schedule[Any, Delay] =
```
Try to return `Duration` here instead of `Delay`, by running the `Delay` (`Duration` is much easier to work with, and has comparison, equals, hash code, etc.). We might need to add `mapM` to schedule to do that easily.

jdegoes commented on core/shared/src/main/scala/scalaz/zio/ZSchedule.scala:
```diff
@@ -654,33 +665,35 @@ trait Schedule_Functions extends Serializable {
    */
-  final def exponential(base: Duration, factor: Double = 2.0): Schedule[Any, Duration] =
-    delayed(forever.map(i => base * math.pow(factor, i.doubleValue)))
+  final def exponential(base: Duration, factor: Double = 2.0): Schedule[Any, Delay] =
```
Try to return `Duration` here instead of `Delay`, by running the `Delay` (`Duration` is much easier to work with, and has comparison, equals, hash code, etc.). We might need to add `mapM` to schedule to do that easily.

alexandredantas commented:
@jdegoes I'm still on this PR. Just got some issues on my job but definitely will fix your requested changes

jdegoes commented:
@alexandredantas No worries, let me know if you need any help!

alexandredantas commented:
@jdegoes Applied your changes the best I could. I didn't write a test for schedule yet because never wrote a property based test before. Can you give me some tips on how to do that?

jdegoes commented on core/shared/src/main/scala/scalaz/zio/ZIO.scala:
```diff
@@ -839,15 +841,73 @@ sealed trait ZIO[-R, +E, +A] extends Serializable { self =>
+   * Schedule this effect using the given schedule.
+   * Scheduled effects will run only when schedule signal it can run, keeping it dormant while it can't or until it fails.
+   */
+  final def scheduled[R1 <: R, B](schedule: ZSchedule[R1, Unit, B]): ZIO[R1 with Clock, E, B] =
```
How do these functions differ from `repeat` etc.?

alexandredantas replied:
Repeat and retry have a semantic of running effect first before evaluating schedule. Scheduled instead evaluates first before running effect, possibly blocking operation since the beginning. Does it make sense?

jdegoes replied:
Maybe we should change `repeat` semantic to be `schedule`?

alexandredantas replied:
Good question :smile: 
I can see both being used

jdegoes replied:
Here's my view:

If we change `repeat` to have semantic of `schedule`, then if you want _old_ behavior, you can just do: `effect *> effect.repeat(schedule)`.

Right?

If this is correct, it's a very good argument that new `schedule` is more foundational than the old `repeat`.

alexandredantas replied:
Makes a lot of sense!

I need to get used to see things as compositional, instead offering a hard method with a hard functionality

I'm really learning a lot with this PR! :smiley:

alexandredantas replied:
Just stumbled in a wall and want some help :)
Scheduled signature is:
`final def scheduled[R1 <: R, B](schedule: ZSchedule[R1, Unit, B]): ZIO[R1 with Clock, E, B]`

It means it receives only schedules doesn't need any external input to calculate next state. It's a bit restrictive, right?

But without Unit, I can't see how to block since the beginning.

How can i solve this?

jdegoes commented on core/shared/src/main/scala/scalaz/zio/ZSchedule.scala:
```diff
@@ -98,6 +103,20 @@ trait ZSchedule[-R, -A, +B] extends Serializable { self =>
+  /**
+   * Returns a new schedule that effectfully mapping over the outputs of this one.
+   */
+  final def mapM[A1 <: A, C](f: B => UIO[C]): ZSchedule[R, A1, C] =
```
Looks good, super handy to have this one!

jdegoes commented on core/shared/src/main/scala/scalaz/zio/ZSchedule.scala:
```diff
@@ -637,8 +664,8 @@ private[zio] trait Schedule_Functions extends Serializable {
    */
-  final val delay: Schedule[Any, Duration] =
-    forever.reconsider[Any, Duration]((_, d) => d.copy(finish = () => d.delay))
+  final val delay: Schedule[Any, Delay] =
```
We should probably add a `final val duration: Schedule[Any, Duration]` that uses `mapM` on `delay` to convert the `Delay` into a `Duration` since those are vastly easier to work with.

alexandredantas replied:
@jdegoes I've added `duration` implementation as:

`final val duration: Schedule[Any, Duration] = delay.mapM(_.run)`

However, when testing this with `>>>`, I was always getting a zeroed `Delay`. Even testing with master branch I'm getting the same result.

When debugging, I found that `delay.reconsider` is happening before `>>>.combineWith`, always issuing a zeroed delay.

jdegoes replied:
This is my mistake. In order to make it work with `>>>` and the current type signature, we'd have to change `update` to take a "preceding" delay (basically threading it through the composition). 

Now, this change may independently be a good change. But I think it's a mistake to conflate it with this refactoring, we can always do it separately.

So what I would suggest is change `duration` to have this type:

```scala
final val duration: Schedule[Delay, Duration] = 
```

Then it can operate on the input to the `Schedule`, which will be a `Delay`.

We already have `fromFunction`, if we add `fromFunctionM`, then we can implement `duration` with that, and add a new useful combinator in the process.

Does all this make sense?

alexandredantas replied:
I've come with this:

``` 
final val duration: Schedule[Delay, Duration] =
    forever.reconsider[Delay, Delay]((s, d) => d.rightMap(_ => s)).mapM(_.run)
```

Does it make sense?

jdegoes replied:
This looks good to me! 😄

jdegoes commented on core/shared/src/main/scala/scalaz/zio/ZSchedule.scala:
```diff
@@ -671,33 +698,43 @@ private[zio] trait Schedule_Functions extends Serializable {
+  final def fibonacci(one: Duration): Schedule[Any, Duration] = {
+    val f = delayed[Any, Any](
+      unfold[(Delay, Delay)]((Delay.none, one.relative)) {
+        case (a1, a2) => (a2, a1 + a2)
```
Uh oh. The `+` operator here will build `Delay` of increasing size, and eventually consume all memory (!). We will have to unfold using `Duration` and map to `Delay`.

jdegoes commented on core/shared/src/main/scala/scalaz/zio/ZSchedule.scala:
```diff
@@ -671,33 +698,43 @@ private[zio] trait Schedule_Functions extends Serializable {
-    delayed(forever.map(i => base * i.doubleValue()))
+  final def linear(base: Duration): Schedule[Any, Duration] = {
+    val l = delayed[Any, Any](forever.map(i => (base * i.doubleValue()).relative))
+    l.mapM(_.run) // TODO: Dotty doesn't infer this properly
```
This pattern is so common, because of the type inference problems, I suggest that we use `l >>> duration` (with the above definition of `duration` that I suggested), both here and elsewhere. Bonus: Dotty may like it more than `mapM`, so we can eliminate the temporary variables and manual type annotations.

alexandredantas replied:
Just tried this approach an even with `l >>> duration` dotty complained about missing implicit for `ConformsR`. With manual type annotations things went fine

jdegoes commented on core/shared/src/main/scala/scalaz/zio/ZSchedule.scala:
```diff
@@ -721,18 +758,22 @@ object ZSchedule extends Schedule_Functions {
   implicit val ConformsAnyProof: ConformsR[Any] = _ConformsR1
 
-  sealed case class Decision[+A, +B](cont: Boolean, delay: Duration, state: A, finish: () => B) { self =>
+  sealed case class Decision[+A, +B](cont: Boolean, delay: Delay, state: A, finish: () => B) {
```
This should probably be `final case class` rather than `sealed case class` (I know you didn't change this, I just saw it 😄 ).

alexandredantas replied:
No problem, I'll change it :)

jdegoes commented on core/shared/src/main/scala/scalaz/zio/ZSchedule.scala:
```diff
@@ -790,9 +834,33 @@ object ZSchedule extends Schedule_Functions {
+  /**
+   * Builds an Schedule capable of running an effect at a given minute and hour, every day
+   */
+  final def everyDayAt(minute: Int, hour: Int): ZSchedule[Clock, Unit, (Long, Long)] =
```
This is _SUPER_ helpful to see in here, as a blueprint for people to make their own.

I'd rename to `everyDay`. I think there's a DSL hiding in here somewhere but let's not worry about that in this PR.

jdegoes commented on core/shared/src/main/scala/scalaz/zio/ZSchedule.scala:
```diff
@@ -790,9 +834,33 @@ object ZSchedule extends Schedule_Functions {
+  final def everyDayAt(minute: Int, hour: Int): ZSchedule[Clock, Unit, (Long, Long)] =
+    ZSchedule[Clock, Long, Unit, (Long, Long)](
+      initial0 = clock.currentTime(unit = TimeUnit.MILLISECONDS).map(_ => 0L),
+      update0 = (_, timesRan) =>
```
We should at least document the dependency on java.time in the microsite, on the Scala.js page.

jdegoes commented on core/shared/src/main/scala/scalaz/zio/delay/Delay.scala:
```diff
@@ -0,0 +1,81 @@
+  /**
+   * Computes delay related to current time in milliseconds
+   */
+  def run: ZIO[Any, Nothing, Duration] = currentTime(unit = TimeUnit.MILLISECONDS).provide(Clock.Live).flatMap {
```
`final`

jdegoes commented on core/shared/src/main/scala/scalaz/zio/duration/Duration.scala:
```diff
@@ -60,6 +63,9 @@ sealed trait Duration extends Ordered[Duration] with Serializable with Product {
   def asJava: JavaDuration
+
+  final def relative: Delay = Delay.relative(this)
+  final def absolute: Delay = Delay.absolute(toMillis)
```
Does `absolute` really make sense on `Duration`???

alexandredantas replied:
You're right. Doesn't make sense. I'll remove it

jdegoes commented on microsite/src/main/tut/datatypes/schedule.md:
````diff
@@ -78,7 +79,7 @@ val jitteredExp = Schedule.exponential(10.milliseconds).jittered
 
 ```scala mdoc:silent
-val boosted = Schedule.spaced(1.second).delayed(_ + 100.milliseconds)
+val boosted = Schedule.spaced(1.second).delayed(_ + 100.milliseconds.relative)
````
All these `.relative` make me want the "syntax" on `Delay`, e.g. `100.millis` could give you a `Delay`. If we don't get something like that, then this will be a source breaking change for existing users. Let me think about that for a bit.

jdegoes commented:
@alexandredantas This is great work! Looking forward to having this capability in core.

jdegoes commented:
@AL333Z This is looking great to me!

My only outstanding concern is if there is a way to preserve backward compatibility (right now, users will have to do `xxx.relative` wherever they were using `xxx`; with a type class, we may be able to eliminate that). Give me a bit and I'll come back with an idea here.

regiskuckaertz commented:
> users will have to do `xxx.relative` wherever they were using `xxx`

also `100.ms.absolute` will never make sense, will it? it would be super nice to have a panoply of combinators for future points in time: `tomorrow.at(6.am)`, `monday.in(2.week)`, etc. 😃

jdegoes commented on core/shared/src/main/scala/scalaz/zio/delay/Delay.scala:
```diff
@@ -0,0 +1,81 @@
+  /**
+   * Computes delay related to current time in milliseconds
+   */
+  final def run: ZIO[Any, Nothing, Duration] = currentTime(unit = TimeUnit.MILLISECONDS).provide(Clock.Live).flatMap {
```
This `run` should return `ZIO[Clock, Nothing, Duration]`, and get the current time from the clock (not from `Clock.Live`), or else it won't be testable.

jdegoes commented on core/shared/src/main/scala/scalaz/zio/delay/Delay.scala:
```diff
@@ -0,0 +1,81 @@
+ * Delays can be relative to current time (a given duration + current time) or an absolute point in time
+ * Delays make waiting less cpu intensive and avoids clock drifts caused by context switching or hibernation
+ */
+sealed trait Delay extends Serializable with Product { self =>
```
`sealed trait` should be replaced by `sealed abstract class`.

alexandredantas replied:
Just for curiosity sake: what's the difference between them? I still don't get some optimizations related to traits/abstract classes

regiskuckaertz replied:
I believe the reason is the bytecode generated is more efficient, due to how methods are ordered in the virtual table. With abstract classes, the order is predictable and consistent, which can be optimised (I think this leads to `invokevirtual` calls being inlined, but don't quote me on this 😅); that is not the case with traits, and the bytecode (`invokeinterface`) cannot be optimised.

jdegoes replied:
Here's another reason:

<https://typelevel.org/blog/2018/05/09/product-with-serializable.html>

Speaking of which we should test extending `Any` with the modules...

jdegoes commented on core/shared/src/main/scala/scalaz/zio/ZSchedule.scala:
```diff
@@ -671,33 +707,36 @@ private[zio] trait Schedule_Functions extends Serializable {
-    delayed(unfold[(Duration, Duration)]((Duration.Zero, one)) {
-      case (a1, a2) => (a2, a1 + a2)
-    }.map(_._1))
+    delayed[Any, Any](
```
Is the `[Any, Any]` type annotation required? It looks like it was not required in the old version.

alexandredantas replied:
I'm trying running this locally over dotty 0.13.0-RC1 and without type annotation an error is thrown.
```
no implicit argument of type Schedule_Functions.this.ConformsR[Nothing] was found for parameter evidence$2 of method delayed in trait Schedule_Functions
```

jdegoes replied:
Dangit.

jdegoes commented on core/shared/src/main/scala/scalaz/zio/ZIO.scala:
```diff
@@ -839,15 +797,75 @@ sealed trait ZIO[-R, +E, +A] extends Serializable { self =>
+  final def scheduleOrElseEither[R1 <: R, B, E2, C](
+    schedule: ZSchedule[R1, Option[A], B],
+    orElse: (E, Option[B]) => ZIO[R1 with Clock, E2, C]
+  ): ZIO[R1 with Clock, E2, Either[C, B]] = {
```
The only (non-semantical) difference between these functions and the `repeat*` functions is the `Option[A]` that is fed into the `Schedule`. The `Option[A]` is currently required because you need to get the "first delay" out of the `Schedule`.

However, I think with a small change to `Schedule`, we can make your new semantic the default, and restore the original type as `A` instead of `Option[A]`.

How?

By having `Schedule`'s `initial` state return a _tuple:

```scala
val initial: ZIO[R, Nothing, (Delay, State)]
```

Now as soon as the initial state is computed, the initial delay ("first delay") is also computed; so schedules now have a chance to delay the first execution of the effect.

With this change, we can remove the `schedule*` methods, and change the `repeat*` methods to have your new semantic: they will first respect an initial delay, rather than running _right away_.

Do you think this could work?

alexandredantas replied:
Yup, definitiely I think it work. Already started implementation and for now it's very promising. Do you think I need to replicate this behavior to retry as well or not?

jdegoes replied:
Yes, I think the correct behavior for retry will be: try right away, if it fails, THEN compute initial state, wait first delay, and THEN feed failure into update and begin/continue loop.

We always have to respect initial delay and always feed all A into schedule without skipping any.

alexandredantas commented:
> > users will have to do `xxx.relative` wherever they were using `xxx`
> 
> also `100.ms.absolute` will never make sense, will it? it would be super nice to have a panoply of combinators for future points in time: `tomorrow.at(6.am)`, `monday.in(2.week)`, etc.

@regiskuckaertz you're right: duration.absolute doesn't make sense. I'll remove it (i think i already did) and replace with some smart construction as you proposed, allowing rich combinations. I'm intending to release some of them in this PR but not a full-feature commbination. What do you think?

regiskuckaertz commented:
> some smart construction as you proposed, allowing rich combinations.

I think YASSS! After re-reading, note that all my examples are relative too 😭 I still believe they'd be useful (I've been playing with a `zio-cron` library over schedules), the difficulty is to come up with a vocabulary that makes sense and is composable. But for absolute points in time we might as well ... parse strings, as in: `absolute"2021-01-25"`.

alexandredantas commented:
> > some smart construction as you proposed, allowing rich combinations.
> 
> I think YASSS! After re-reading, note that all my examples are relative too  I still believe they'd be useful (I've been playing with a `zio-cron` library over schedules), the difficulty is to come up with a vocabulary that makes sense and is composable. But for absolute points in time we might as well ... parse strings, as in: `absolute"2021-01-25"`.

I think your examples are very good actually. I see them as absolute because they point to a future moment. `tomorrow.at(6.pm)` for example can be seen as an absolute point in future running only once.

Every absolute point in time can be seen as relative but the structure itself is abstracting delay calculation from programmers :smiley: 

And another upside of introducing it: I was trying to find an way to always create `Absolute` using milliseconds representation. Using duration is confusing and using long can be anything. Introducing some expression structure allows creating absolute points in time hiding complexity from users.

@jdegoes do you think this is the DSL you said that could be hiding on `everyDay` schedule?

jdegoes commented:
> @jdegoes do you think this is the DSL you said that could be hiding on everyDay schedule?

I think the core of the DSL involves adding a `apply` method to `Schedule`, which takes an inner `Schedule`, with the following semantic:

*The outer schedule runs the inner schedule until completion, and then once completed, it takes a step itself, and on the next step, it resets the inner schedule (using its starting state) and runs the inner schedule again until completion, and so forth, until the outer schedule is done.*

Then if we have an `every` method, I think we can do things like:

```scala
every(Monday) {
  every(Hour) {
    every(Minute)
  }
|
```

Then if we have an `after` method, we can combine them:

```scala
every(Monday) {
  after(6.pm) {
    every(10.minutes)
  }
}
```

The difference between `every` and `after` would be `every` keeps doing it on the specified relative schedule, whereas `after` would do it once after the specified elapsed interval.

Maybe others are necessary too, I am not sure, and maybe `.apply` is not a good name.

alexandredantas commented:
> > @jdegoes do you think this is the DSL you said that could be hiding on everyDay schedule?
> 
> I think the core of the DSL involves adding a `apply` method to `Schedule`, which takes an inner `Schedule`, with the following semantic:
> 
> _The outer schedule runs the inner schedule until completion, and then once completed, it takes a step itself, and on the next step, it resets the inner schedule (using its starting state) and runs the inner schedule again until completion, and so forth, until the outer schedule is done._
> 
> Then if we have an `every` method, I think we can do things like:
> 
> ```scala
> every(Monday) {
>   every(Hour) {
>     every(Minute)
>   }
> |
> ```
> 
> Then if we have an `after` method, we can combine them:
> 
> ```scala
> every(Monday) {
>   after(6.pm) {
>     every(10.minutes)
>   }
> }
> ```
> 
> The difference between `every` and `after` would be `every` keeps doing it on the specified relative schedule, whereas `after` would do it once after the specified elapsed interval.
> 
> Maybe others are necessary too, I am not sure, and maybe `.apply` is not a good name.

But if inner schedule runs until completion, some combinations may runs over outer constraints, for example:

every(monday){
  every(minute){
    ...
  }
}

Inner schedule will start running monday but will continue to tuesday, wednesday, etc. Maybe outer schedules should take a step itself, delaying execution if required, then passing control to inner schedule. What do you think?

alexandredantas commented:
I'm placing a very initial draft of schedule expressions to be used as smart constructors for points in time. Please, @regiskuckaertz  and @jdegoes give some tips if i'm going on right direction. My main idea was constraining `Delay.absolute` constructor to use always an expression, avoiding misuse. Don't know if it's the right way to go

regiskuckaertz commented:
> Inner schedule will start running monday but will continue to tuesday, wednesday, etc.

Probably because `minute` is not specific enough and there needs to be `minuteOfTheDay`, `minuteOfTheHour`. Then it is clear the schedule will eventually stop.

regiskuckaertz commented:
> Probably because minute is not specific enough and there needs to be minuteOfTheDay, minuteOfTheHour. Then it is clear the schedule will eventually stop.

... on second thought, the output of `every { every { ... } }` should probably be that of all underlying schedules `&&`-ed. Then the problem disappears as `Monday` will stop as soon as it's Tuesday.

alexandredantas commented:
> > Probably because minute is not specific enough and there needs to be minuteOfTheDay, minuteOfTheHour. Then it is clear the schedule will eventually stop.
> 
> ... on second thought, the output of `every { every { ... } }` should probably be that of all underlying schedules `&&`-ed. Then the problem disappears as `Monday` will stop as soon as it's Tuesday.

I was thinking exactly about that. Nested `every` can be represented as schedule composition, using &&, || and other operators. This way we don't need new structures and can build complex scheduling logic. I think some new constructors like `every` and `after` can help to hide delay calculation from programmers

jdegoes commented:
> I'm placing a very initial draft of schedule expressions to be used as smart constructors for points in time. Please, @regiskuckaertz and @jdegoes give some tips if i'm going on right direction. My main idea was constraining Delay.absolute constructor to use always an expression, avoiding misuse. Don't know if it's the right way to go

Personally, I'd defer "schedule expressions" for a followup PR, because it would be nice to get this one merged in, and then take a little more time to see if / what we need for schedule expressions. We need a list of use cases to solve, e.g.:

 - create a schedule that repeats the fourth Monday every month, and the second Tuesday, and on the second Tuesday, it repeats every 3 hours, at 15 minutes, 30 minutes, and 50 minutes past the hour mark.

This can motivate the combinators and inform the design.

jdegoes commented:
I think what we miss right now (this is just a guess) is a combinator for: AFTER this schedule steps, THEN run this _next_ schedule until it's done, then continue with the original schedule.

I can't see a way to do that from existing combinators. There's some nesting, it's like a nested for loop:

```
for (each step in top schedule) {
  for (each step in lower schedule) {
    ...
  }
}
```

alexandredantas commented:
> > I'm placing a very initial draft of schedule expressions to be used as smart constructors for points in time. Please, @regiskuckaertz and @jdegoes give some tips if i'm going on right direction. My main idea was constraining Delay.absolute constructor to use always an expression, avoiding misuse. Don't know if it's the right way to go
> 
> Personally, I'd defer "schedule expressions" for a followup PR, because it would be nice to get this one merged in, and then take a little more time to see if / what we need for schedule expressions. We need a list of use cases to solve, e.g.:
> 
> * create a schedule that repeats the fourth Monday every month, and the second Tuesday, and on the second Tuesday, it repeats every 3 hours, at 15 minutes, 30 minutes, and 50 minutes past the hour mark.
> 
> This can motivate the combinators and inform the design.

I'd like to introduce at least `EveryDay` expression. It will help build `Absolute` delay without relying on `Duration` or a `Long` value. Does it make sense?

jdegoes commented:
> I'd like to introduce at least EveryDay expression. It will help build Absolute delay without relying on Duration or a Long value. Does it make sense?

I feel that `EveryDay` is a combination of three schedules:

 - A daily repetition
 - Appended to or nested within that, a single (non-repeating) delay of `hour` hours
 - Appended to or nested with that, a single (non-repeating) delay of `minute` minutes

We don't have a way to fuse durations in this way, or to do the nesting (although nesting is not required here, I don't think, because the latter two don't repeat in this specific example).

Actually that's not quite true, we have `>>>`, which may be able to do simple things:

```scala
(day(1) >>> hour(3) >>> minute(3)).forever
```

I think we're missing `schedule1.repeat(schedule2)` (which would allow implementation of `.forever` as `schedule1.repeat(Schedule.forever)`, and maybe that's all.

We're close but I'd hate to ship `EveryDay` if we can break it down into little pieces...

jdegoes commented on core/jvm/src/test/scala/scalaz/zio/QueueSpec.scala:
```diff
@@ -152,16 +152,16 @@ class QueueSpec(implicit ee: org.specs2.concurrent.ExecutionEnv) extends TestRun
     unsafeRun((for {
       queue        <- Queue.bounded[Int](10)
-      _            <- queue.offer(1).repeat(ZSchedule.recurs(9))
+      _            <- queue.offer(1).repeat(ZSchedule.recurs(9).immediately)
```
All the `immediately` everywhere concern me... why do we need to ignore the initial delay?

alexandredantas replied:
Actually `immediately` is giving an initial delay of `zero` to an existent schedule. I thought about it to avoid breaking current schedule implementations.

jdegoes replied:
I think we should encourage the. new pattern in the ZIO code base, i.e. do not run the first one until it's specified to run. 

What does `retry` do with the initial delay? I suppose it would ignore it, right?

alexandredantas replied:
I agree. I just got shy to change signatures :smile: 
But make sense changing tthem

On my last commit I made `retry` considering initial delay like `repeat`

alexandredantas commented:
> > I'd like to introduce at least EveryDay expression. It will help build Absolute delay without relying on Duration or a Long value. Does it make sense?
> 
> I feel that `EveryDay` is a combination of three schedules:
> 
> * A daily repetition
> * Appended to or nested within that, a single (non-repeating) delay of `hour` hours
> * Appended to or nested with that, a single (non-repeating) delay of `minute` minutes
> 
> We don't have a way to fuse durations in this way, or to do the nesting (although nesting is not required here, I don't think, because the latter two don't repeat in this specific example).
> 
> Actually that's not quite true, we have `>>>`, which may be able to do simple things:
> 
> ```scala
> (day(1) >>> hour(3) >>> minute(3)).forever
> ```
> 
> I think we're missing `schedule1.repeat(schedule2)` (which would allow implementation of `.forever` as `schedule1.repeat(Schedule.forever)`, and maybe that's all.
> 
> We're close but I'd hate to ship `EveryDay` if we can break it down into little pieces...

Ok. So for now I'll continue to build `Absolute` using a long representing epoch in milliseconds.

jdegoes commented:
> About the composition of EveryDay, I totally agree. I just don't know how to implement it for now 😃
> Maybe I should leave it for another PR. What do you think?

Yes, let's do that! Excited to get this one in for the next RC.

alexandredantas commented:
> While you are wrapping this up, I thought of something for `retry`:
> 
> Retry will invoke the schedule in case of a failure. It may make sense to invoke the schedule right away, and add the initial delay with whatever is returned by the schedule.
> 
> In addition, we should make sure that the initial delays compose using `&&` and `||` and `composeWith`, appropriately. Just like the "step" delays are combined, the initial delays should be combined in the same way.
> 
> Finally, we should make the stream repetition semantics match the ordinary one.
> 
> Let me know if this makes sense or if you disagree with any of this.

I agree. Actually, I think this will be the base for that schedule dsl we've discussed. `every` and `at` could be operators like `&&` and `||`, allowing combining initial delays and evaluating if schedules should start.

I'll push this change soon

alexandredantas commented:
@jdegoes "Retry will invoke the schedule in case of a failure. It may make sense to invoke the schedule right away, and add the initial delay with whatever is returned by the schedule."
About this: schedule will return return delay for next step. Does it make sense to add initial delay to step delay?

jdegoes commented:
> About this: schedule will return return delay for next step. Does it make sense to add initial delay to step delay?

It seems like we either add them, or throw away the initial delay. It doesn't make sense (in the case of retry) to invoke the action, get a failure, wait the initial delay, pass the error to the schedule, and then get the next delay, and wait that. It will introduce a spurious wait between the time of the failure, and the time when the schedule knows about the failure. 

Does that make sense? What do you think?

alexandredantas commented:
Initial delay is being considered before executing effect for the first time (before foldM). After that, I think delay logic responsability should be passed to `update` only, as it is currently. I don't know if make sense adding initial delay here as we already have `update` calculating the next delay. Maybe we could pass initial delay to `update` so it could be taken in. What do you think?

jdegoes commented:
> Initial delay is being considered before executing effect for the first time (before foldM). After that, I think delay logic responsability should be passed to update only, as it is currently. 

I think the advantage of doing it this way is that it's consistent with `repeat`.

In practice, I don't think user will want to wait before executing an effect, if they only want to retry it if it fails. They may be surprised to change code like `makeRequest` into `makeRequest.retry(schedule)`, and then find out the request doesn't happen right away, because of an initial delay.

However, there's another way to solve that: just making sure only the date/time combinators use the initial delay. It can be 0 for all others. In fact, before we sleep, we should check, and if it's 0, we should skip the sleep (that will result in less latency than the current approach).

Do you think that could work?

alexandredantas commented:
> > Initial delay is being considered before executing effect for the first time (before foldM). After that, I think delay logic responsability should be passed to update only, as it is currently.
> 
> I think the advantage of doing it this way is that it's consistent with `repeat`.
> 
> In practice, I don't think user will want to wait before executing an effect, if they only want to retry it if it fails. They may be surprised to change code like `makeRequest` into `makeRequest.retry(schedule)`, and then find out the request doesn't happen right away, because of an initial delay.
> 
> However, there's another way to solve that: just making sure only the date/time combinators use the initial delay. It can be 0 for all others. In fact, before we sleep, we should check, and if it's 0, we should skip the sleep (that will result in less latency than the current approach).
> 
> Do you think that could work?

About retry: i think you're right.

I didn't get the part of date/time combinators using initial delay. Do you mean only considering initial delay on `Schedule.update`?

jdegoes commented:
@alexandredantas Do you want to have a Google Hangout session? I think it would help us to wrap up the final issues.

alexandredantas commented:
> @alexandredantas Do you want to have a Google Hangout session? I think it would help us to wrap up the final issues.

Sounds great!! I'd just finished merging my code with organization changes.

alexandredantas commented:
@jdegoes when do you want to schedule the hangout session?

alexandredantas commented:
@jdegoes any update? :)

jdegoes commented:
@alexandredantas How about now? 😄 

Did you get everything passing???

alexandredantas commented:
@jdegoes sorry for not getting the message early. what about friday? anytime :)

About the tests, yes, all passing!! I've merged with master and looks like something was fixed

jdegoes commented:
This work is being continued in #1798. @alexandredantas Thank you for all your help on this and if you have some time to contribute, your assistance reviewing #1798 and ensuring we satisfy all requirements of the original ticket would be greatly appreciated. 🙏
